include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -294,6 +294,37 @@ def ConstantOp : MooreOp<"constant", [Pure]> {
   ];
 }
 
+def Parameter : I32EnumAttrCase<"Parameter", 0, "parameter">;
+def LocalParameter : I32EnumAttrCase<"LocalParameter", 1, "localparam">;
+def SpecParameter : I32EnumAttrCase<"SpecParameter", 2, "specparam">;
+
+def NamedConstAttr : I32EnumAttr<"NamedConst", "elaboration-time constants", 
+       [Parameter, LocalParameter, SpecParameter]>{
+    let cppNamespace = "circt::moore";
+}
+
+def NamedConstantOp : MooreOp<"named_constant", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+  OptionalTypesMatchWith<"value and named constant types match",
+    "result", "value", "$_self">,
+]>{
```

Comment:
The `OptionalTypesMatchWith` is different from `TypesMatchWith`. It's just used to either `lhs` or `rhs` are optional. `SameOperandsAndResultType` is a better choice.

Reply:
👍 for `SameOperandsAndResultType` 😃

Reply:
Good idea! I will add `SameOperandsAndResultType` for it.